Adds: dry-run, sync subset, and Fig autocomplete

jazanne opened the pull request:
## Summary

### Adds new options

- `--dry-run` - preview changes
- `--verbose` - log output setting
- `-f`, `--flag` - sync one flag
- `-T`,  `--tag` - sync flags with tag(s)

### Adds support for [Fig autocomplete](https://fig.io/)

Note, some options like `isRepeatable` work for `--projects`, but not `-p`. I kept them since it is presumably a bug that will be fixed.


### Support ldc config

- use a config file (same format as LaunchDarkly-Labs/ldc) to easily autocomplete api tokens and app uri

### Fixes

- Errors not thrown correctly for non-existent environments

## Testing

Download [Fig](https://fig.io/)

Navigate to this repo/branch in terminal

Run script with `./sync-ld-flags` to get Fig autocompletions

_See README for more details on using script_

mmrj commented on README.md:
````diff
@@ -20,21 +19,95 @@ The following properties will be synchronized:
-        node index.js -p PROJECT_KEY -s SOURCE_ENVIRONMENT_KEY -d DEST_ENVIRONMENT_KEY -t API_TOKEN -H https://your-launch-darkly-deploy.com
+   ```
+   ./sync-ld-flags --api-token API_TOKEN \
+     -H https://your-launch-darkly-deploy.com
````
Will the user need single quotes around the URL here?

jazanne replied:
nope, should be ok without it

mmrj commented on README.md:
````diff
@@ -20,21 +19,95 @@ The following properties will be synchronized:
+
+```
+  -p, --project-key <key>      Project key
+  -s, --source-env <key>       Source environment
````
```suggestion
  -s, --source-env <key>       Source environment key
```

mmrj commented on README.md:
```diff
@@ -20,21 +19,95 @@ The following properties will be synchronized:
+
+![Auto-complete configuration](docs/images/autocomplete-config.gif)
+
+Auto-complete keys and servers by using a config file:
+
+Add the following configuration to `~/.config/ldc.json`
```
```suggestion
You can autocomplete keys and servers by using a config file. Add the following configuration to `~/.config/ldc.json`:
```

mmrj commented on README.md:
```diff
@@ -20,21 +19,95 @@ The following properties will be synchronized:
+
+Want to set these configuration variables more easily?
+
+![Auto-complete configuration](docs/images/autocomplete-config.gif)
```
```suggestion
![Autocomplete configuration.](docs/images/autocomplete-config.gif)
```

jazanne commented on example-config.json:
```diff
@@ -0,0 +1,6 @@
+        "apitoken": "<REPLACE WITH PERSONAL ACCESS TOKEN>",
+        "server": "<REMOVE OR REPLACE WITH APP URI>"
+    }
+}
```
```suggestion
}

```

yusinto commented on index.js:
```diff
@@ -7,6 +7,9 @@ const jsonpatch = require('fast-json-patch');
 const request = require('request');
```
I think casing consistency matters so in this case should `jsonpatch` be `jsonPatch` instead?

jazanne replied:
hum i don't think it makes a difference since we don't use the latter anywhere that i'm aware of and the module itself uses all lowercase in examples https://www.npmjs.com/package/fast-json-patch

jazanne replied:
AH! ok that makes sense - i can definitely change that

yusinto commented on index.js:
```diff
@@ -7,6 +7,9 @@ const jsonpatch = require('fast-json-patch');
 const request = require('request');
 const program = require('commander');
 
+// Use to calculcate changing flags
```
Typo `calculcate` => `calculate`

yusinto commented on index.js:
```diff
@@ -28,20 +31,35 @@ function patchFlag(patch, key, config, cb) {
+
   const requestOptions = {
-    url: `${baseUrl}/flags/${projectKey}?summary=0&env=${sourceEnvironment}&env=${destinationEnvironment}`,
+    url: url,
```
This can be simplified to:
```ts
const requestOptions = {
    url,
```

yusinto commented on index.js:
```diff
@@ -28,20 +31,35 @@ function patchFlag(patch, key, config, cb) {
 const fetchFlags = function (config, cb) {
-  const { baseUrl, projectKey, sourceEnvironment, destinationEnvironment, apiToken } = config;
+  const { baseUrl, projectKey, sourceEnvironment, destinationEnvironment, apiToken, tags, flag } = config;
+  let isSingle = flag ? true : false;
```
This can be simplified to:

```ts
let isSingle = !!flag
```

jazanne replied:
i modified this a but to make sure tags aren't included (which they shouldn't be)

yusinto commented on index.js:
```diff
@@ -28,20 +31,35 @@ function patchFlag(patch, key, config, cb) {
+  let isSingle = flag ? true : false;
+  let url = isSingle
+    ? `${baseUrl}/flags/${projectKey}/${flag}?summary=0&env=${sourceEnvironment}&env=${destinationEnvironment}`
+    : `${baseUrl}/flags/${projectKey}?summary=0&env=${sourceEnvironment}&env=${destinationEnvironment}${tags ? ("&filter=tags:" + tags.join('+')) : '' }`;
```
Nested ternaries are quite difficult to read and debug so much so [eslint has a rule disallowing it](https://eslint.org/docs/rules/no-nested-ternary). I recommend we avoid nesting ternaries like this and be more verbose using if/else for example.

jazanne replied:
good call - this was hard to follow

yusinto commented on .fig/autocomplete/src/sync-ld-flags.ts:
```diff
@@ -0,0 +1,283 @@
+      },
+    },
+    {
+      name: ["--dry-run"],
```
Any reason why there's no shortcode for `dry-run`?

jazanne replied:
i wasn't sure what a good shortcode would be since we already use `-d` and `-D` for other opts and i couldn't find any examples of what people typically use for this - open to suggestions!

yusinto replied:
I looked around, both `make` and `yarn` uses `-n` as an alias for `--dry-run` so we can follow that.

yusinto commented:
I would like to report a crash which I discovered, when I ran the command using the `-f` param with a non-existent flag:

```bash
./sync-ld-flags --dry-run --api-token api-xxx --project-key string --source-env production --destination-env test -v -f nonExistentFlagKey

/Users/yusngadiman/projects/sync-ld-flags-j/index.js:169
      if (matches.groups && matches.groups.envKey) {
                  ^

TypeError: Cannot read properties of null (reading 'groups')
    at /Users/yusngadiman/projects/sync-ld-flags-j/index.js:169:19
    at Request.callback [as _callback] (/Users/yusngadiman/projects/sync-ld-flags-j/index.js:62:7)
    at Request.self.callback (/Users/yusngadiman/projects/sync-ld-flags-j/node_modules/request/request.js:185:22)

```

jazanne commented on .fig/autocomplete/package.json:
```diff
@@ -0,0 +1,71 @@
+{
+  "name": "boilerplate-fig-autocomplete",
+  "version": "1.0.0",
+  "description": "Boilerplate Fig autocomplete spec package",
```
My thought is no, because this is actually a package.json for the [Fig dev tools](https://fig.io/docs/guides/private-autocomplete), not our own package.

open to changing it though if you think we should

jazanne commented:
@yusinto I fixed the unhandled flag error (404) you found

I also added some basic linting since i realized we didn't have it when i made the `jsonPatch` change - lmk what you think. I can also revert and handle in separate PR if you think that's better

yusinto commented:
@jazanne linting is a great idea. Thanks for doing this!
